### PR TITLE
[DPE-5052] COS metrics -- Prometheus, Grafana, Alerting

### PR DESCRIPTION
**Pull request by juditnovak**
Kyuubi integration to COS:
 - Prometheus reporting
 - Grafana
 - Alerting

Grafana Dasboards:
image

Prometheus Dashboard:
image
image

Alerts:
image

image

NOTE: [Kyuubi metrics](https://kyuubi.readthedocs.io/en/master/monitor/metrics.html#metrics) provided out of the box:
image

Integration:
```
curl -L https://raw.githubusercontent.com/canonical/cos-lite-bundle/main/overlays/offers-overlay.yaml -O
curl -L https://raw.githubusercontent.com/canonical/cos-lite-bundle/main/overlays/storage-small-overlay.yaml -O
juju deploy cos-lite --trust --overlay ./offers-overlay.yaml --overlay ./storage-small-overlay.yaml

juju relate grafana-agent-k8s kyuubi-k8s:grafana-dashboard
juju relate grafana-agent-k8s kyuubi-k8s:metrics-endpoint
```

As a result, the traefik operator should start serving on an IP announced in `juju status`:
image

..where the COS Catalouge should be available such as: `http://<traefik_IP>/<juju_model_name>-catalouge`

IF NOT:
Make sure that the following relations are available:
 - `juju relate grafana-agent-k8s:grafana-dashboards-provider grafana`
 - `juju relate grafana-agent-k8s:send-remote-write prometheus`

### CODE REVIEW

File: tests/integration/helpers.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+from constants import COS_METRICS_PORT
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+OPENSEARCH_APP_NAME = "opensearch"
+
+
+def prometheus_exporter_data(host: str) -> str | None:
+    """Check if a given unit has 'kibana-exporter' service available and publishing."""
```
Comment (juditnovak) [inline tests/integration/helpers.py:21]: Note to self: all docstrings to be fixed

File: pyproject.toml [hunk 2]
```diff
@@ -39,6 +40,12 @@ lightkube = "^0.15.2"
 psycopg2 = "^2.9.9"
 spark8t = { git = "https://github.com/canonical/spark-k8s-toolkit-py.git", tag="v0.0.7", python=">3.8,<4.0"}
 
+[tool.poetry.group.charm-libs.dependencies]
+ops = "^2.13.0"
+poetry-core = "^1.9.0"
+# grafana_agent/v0/cos_agent.py requires pydantic <2
+pydantic = "^1.10.17, <2"
```
Comment (deusebio) [inline pyproject.toml:47]: **praise** Nice! I really believe we should add these acrsso the repositories @theoctober19th @welpaolo

**Review by deusebio — APPROVED** [review]
Looks very good! Well done!

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
```
Comment (welpaolo) [inline tests/integration/helpers.py:2]: Please update copyright date

File: tests/integration/helpers.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+
+import requests
+import yaml
+from pytest_operator.plugin import OpsTest
+
+from constants import COS_METRICS_PORT
+
+METADATA = yaml.safe_load(Path("./metadata.yaml").read_text())
+APP_NAME = METADATA["name"]
+OPENSEARCH_APP_NAME = "opensearch"
```
Comment (welpaolo) [inline tests/integration/helpers.py:17]: Please remove `opensearch` mention.

File: tests/integration/helpers.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+
+def prometheus_exporter_data(host: str) -> str | None:
+    """Check if a given host has 'kibana-exporter' service available and publishing."""
```
Comment (welpaolo) [inline tests/integration/helpers.py:21]:
```suggestion
    """Check if a given host has metric service available and it is publishing."""
```

File: tests/integration/helpers.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    try:
+        response = requests.get(url)
+    except requests.exceptions.RequestException:
+        return
+    if response.status_code == 200:
+        return response.text
+
+
+async def all_prometheus_exporters_data(ops_test: OpsTest, check_field) -> bool:
+    """Check if a all units has 'kibana-exporter' service available and publishing."""
```
Comment (welpaolo) [inline tests/integration/helpers.py:32]:
```suggestion
    """Check if a all units has metric service available and publishing."""
```

File: tests/integration/test_charm.py [hunk 7]
```diff
@@ -967,3 +979,86 @@ async def test_read_spark_properties_from_secrets(ops_test: OpsTest, test_pod):
...
     assert set(executor_pod_names) == set(expected_executor_pod_names)
+
+
+@pytest.mark.abort_on_fail
+async def test_kyuubi_cos_relation_joined(ops_test: OpsTest):
+    # Prometheus data is being published by the app
+    assert await all_prometheus_exporters_data(ops_test, check_field="kyuubi_jvm_uptime")
+
+    # Deploying and relating to grafana-agent
+    logger.info("Deployeing grafana-agent-k8s charm...")
```
Comment (welpaolo) [inline tests/integration/test_charm.py:990]:
```suggestion
    logger.info("Deploying grafana-agent-k8s charm...")
```

**Review by welpaolo — APPROVED** [review]
Looks good to me! 
I left some minor comments.

File: tests/integration/helpers.py (hunk 4 above)
Comment (juditnovak) [reply tests/integration/helpers.py:17]: :sweat_smile:

**Pull request merged**